File: 2-7/2-7-TS/Vehicle.ts
/*
 * Class representing a vehicle with basic functionality.
 * @version 1.0
 * @since   2024-11-05
 */

export default class Vehicle {

    /**
     * The color of the vehicle.
     * @type {string}
     */
    protected color: string;

    /**
     * The maximum speed of the vehicle.
     * @type {number}
     */
    protected maxSpeed: number;

    /**
     * The current speed of the vehicle.
     * @type {number}
     */
    protected currentSpeed: number;

    /**
     * Creates a new Vehicle instance.
     * @param {string} color - The color of the vehicle.
     * @param {number} maxSpeed - The maximum speed of the vehicle.
     */
    constructor(color: string, maxSpeed: number) {
        this.color = color;
        this.maxSpeed = maxSpeed;
        this.currentSpeed = 0;
    }

    /**
     * Returns the current speed of the vehicle.
     * @returns {number} The current speed of the vehicle.
     */
    get speed(): number {
        return this.currentSpeed;
    }

    /**
     * Returns the maximum speed of the vehicle.
     * @returns {number} The maximum speed of the vehicle.
     */
    get maxVehicleSpeed(): number {
        return this.maxSpeed;
    }

    /**
     * Accelerates the vehicle by a specified power over a given time.
     * @param {number} power - The power level for acceleration.
     * @param {number} time - The time duration for acceleration.
     */
    public accelerate(power: number, time: number): void {
        const speedIncrease = power * time;
        if (this.currentSpeed + speedIncrease < this.maxSpeed) {
            this.currentSpeed += speedIncrease;
        } else {
            this.currentSpeed = this.maxSpeed;
        }
    }

    /**
     * Brakes the vehicle by a specified power over a given time.
     * @param {number} brakePower - The power level for braking.
     * @param {number} brakeTime - The time duration for braking.
     */
    public brake(brakePower: number, brakeTime: number): void {
        const speedDecrease = brakePower * brakeTime;
        this.currentSpeed = Math.max(this.currentSpeed - speedDecrease, 0);
    }

    /**
     * Displays the status of the vehicle.
     */
    public status(): void {
        console.log(`-> Speed: ${this.currentSpeed}`);
        console.log(`-> Max Speed: ${this.maxSpeed}`);
        console.log(`-> Color: ${this.color}`);
    }
}

File: 2-I0/2-10-ruby/vehicle.rb
# frozen_string_literal: true

# Class representing a vehicle with basic functionality.
# @version 1.0
# @since   2024-11-05

# The Vehicle class which represents a vehicle
class Abstract Vehicle
  # The color of the vehicle.
  # @type [String]
  attr_reader :color

  # The maximum speed of the vehicle.
  # @type [Integer]
  attr_reader :max_speed

  # The current speed of the vehicle.
  # @type [Integer]
  attr_reader :current_speed

  # The number of tires of the vehicle.
  # @type [Integer]
  attr_reader :number_of_tires

  # Creates a new Vehicle instance.
  # @param color [String] The color of the vehicle.
  # @param max_speed [Integer] The maximum speed of the vehicle.
  def initialize(color, max_speed, number_of_tires)
    @color = color
    @max_speed = max_speed
    @current_speed = 0
    @number_of_tires = number_of_tires
  end

  # Returns the current speed of the vehicle.
  # @return [Integer] The current speed of the vehicle.
  def speed
    @current_speed
  end

  # Returns the maximum speed of the vehicle.
  # @return [Integer] The maximum speed of the vehicle.
  def max_vehicle_speed
    @max_speed
  end

  # Returns the number of tires of the vehicle.
  # @return [Integer] The number of tires of the vehicle.
  def vehicle_number_of_tires
    @number_of_tires
  end

  # Accelerates the vehicle by a specified power over a given time.
  # @param power [Integer] The power level for acceleration.
  # @param time [Integer] The time duration for acceleration.
  def accelerate(power, time)
    speed_increase = power * time
    if @current_speed + speed_increase < @max_speed
      @current_speed += speed_increase
    else
      @current_speed = @max_speed
    end
  end

  # Brakes the vehicle by a specified power over a given time.
  # @param power [Integer] The power level for braking.
  # @param time [Integer] The time duration for braking.
  def brake(power, time)
    speed_decrease = power * time
    if (@current_speed - speed_decrease).positive?
      @current_speed -= speed_decrease
    else
      @current_speed = 0
    end
  end

  # Displays the status of the vehicle.
  def status
    puts "-> Speed: #{@current_speed}"
    puts "-> Max Speed: #{@max_speed}"
    puts "-> Color: #{@color}"
    puts "-> # of tires: #{@number_of_tires}"
  end
end
